src/bootstrap/config.rs

```diff
@@ -820,44 +729,25 @@ impl Config {
         config.bindir = "bin".into();
         config.dist_include_mingw_linker = true;
 
-        // set by build.rs
-        config.build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
-
-        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-        // Undo `src/bootstrap`
-        config.src = manifest_dir.parent().unwrap().parent().unwrap().to_owned();
-        config.out = PathBuf::from("build");
-
         config
     }
 
-    pub fn parse(args: &[String]) -> Config {
-        #[cfg(test)]
-        let get_toml = |_: &_| TomlConfig::default();
-        #[cfg(not(test))]
-        let get_toml = |file: &Path| {
-            let contents =
-                t!(fs::read_to_string(file), format!("config file {} not found", file.display()));
-            // Deserialize to Value and then TomlConfig to prevent the Deserialize impl of
-            // TomlConfig and sub types to be monomorphized 5x by toml.
-            match toml::from_str(&contents)
-                .and_then(|table: toml::Value| TomlConfig::deserialize(table))
-            {
-                Ok(table) => table,
-                Err(err) => {
-                    eprintln!("failed to parse TOML configuration '{}': {}", file.display(), err);
-                    crate::detail_exit(2);
-                }
-            }
-        };
-
-        Self::parse_inner(args, get_toml)
-    }
-
-    fn parse_inner<'a>(args: &[String], get_toml: impl 'a + Fn(&Path) -> TomlConfig) -> Config {
+    pub fn parse(args: &[String], custom_toml_config: Option<&str>) -> Config {
         let flags = Flags::parse(&args);
         let mut config = Config::default_opts();
 
+        let mut toml: TomlConfig = if let Some(custom_toml_config) = custom_toml_config {
+            toml::from_str(custom_toml_config).unwrap()
+        } else {
+            set_cfg_path_and_return_toml_cfg(
+                config.src.clone(),
+                flags.config.clone(),
+                &mut config.config,
+            )
+        };
+
+        config.minimal_config = MinimalConfig::parse(&flags, toml.build.clone());
```

Comment:
This feels like the wrong approach to me. We should *always* start with MinimalConfig::parse, and then take that as an argument for the big Config struct. Any processing that is necessary for the minimal config is likely to *also* be necessary here, and can be read from fields in that struct as needed.

I think that will help remove the helper functions shared between the two (e.g., set_cfg_path_and_return_toml_cfg) and make the relationship between the two clearer in general.

Reply:
> We should always start with MinimalConfig::parse, and then take that as an argument for the big Config struct.

Right now `min_config` module is short and simple. If I apply the refactoring based on this suggestion, this will cause having a lot of `Config` field calculations in `min_config` module which will kill the seperation between them. Which will make the maintenance harder for both `Config` and `MinConfig`. Are you sure you want me to do this?

Reply:
I don't understand. Why would Config field calculations need to move into MinConfig for this refactoring? It seems like only the calculations for the fields already in MinConfig would need to move; nothing else should. And those calculations already need to be in MinConfig anyway, as otherwise we can't just parse it?

Reply:
So you want `MinConfig` field mapping to be moved in `config` module then. I am still unsure if this will be better. We intentionally did this seperation with 54b7ff25547f5893932ba0319234206a2d5fe2c4,  but I can do it if you want.